plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/executors/ReactorScheduledExecutorService.java

```diff
@@ -0,0 +1,188 @@
+ */
+
+package org.elasticsearch.repositories.azure.executors;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.SuppressForbidden;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.AbstractExecutorService;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Delayed;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Wrapper around {@link ThreadPool} that provides the necessary scheduling methods for a {@link reactor.core.scheduler.Scheduler} to
+ * function. This allows injecting a custom Executor to the reactor schedulers factory and get fine grained control over the
+ * thread resources used.
+ */
+@SuppressForbidden(reason = "It wraps a ThreadPool and delegates all the work")
+public class ReactorScheduledExecutorService extends AbstractExecutorService implements ScheduledExecutorService {
+    private final ThreadPool threadPool;
+    private final String executorName;
+    private final ExecutorService delegate;
+    private final Logger logger = LogManager.getLogger(ReactorScheduledExecutorService.class);
+
+    public ReactorScheduledExecutorService(ThreadPool threadPool, String executorName) {
+        this.threadPool = threadPool;
+        this.executorName = executorName;
+        this.delegate = threadPool.executor(executorName);
+    }
+
+    @Override
+    public <V> ScheduledFuture<V> schedule(Callable<V> callable, long delay, TimeUnit unit) {
+        Scheduler.ScheduledCancellable schedule = threadPool.schedule(() -> {
+            try {
+                decorateCallable(callable).call();
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }, new TimeValue(delay, unit), executorName);
+
+        return new ReactorFuture<>(schedule);
+    }
+
+    public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
+        Runnable decoratedCommand = decorateRunnable(command);
+        Scheduler.ScheduledCancellable schedule = threadPool.schedule(decoratedCommand, new TimeValue(delay, unit), executorName);
+        return new ReactorFuture<>(schedule);
+    }
+
+    @Override
+    public ScheduledFuture<?> scheduleAtFixedRate(Runnable command, long initialDelay, long period, TimeUnit unit) {
+        Runnable decoratedCommand = decorateRunnable(command);
+
+        return threadPool.scheduler().scheduleAtFixedRate(() -> {
+            try {
+                delegate.execute(decoratedCommand);
```

Comment:
It's possible that the command is executed some time after the specified `period` if the thread pool backed by `delegate` is over subscribed. This is used for timeouts and retries, I guess we could live with a less precise timers but still control the executor in which that code is executed.